core/src/repair_weighted_traversal.rs

```diff
@@ -69,10 +69,119 @@ impl<'a> Iterator for RepairWeightTraversal<'a> {
     }
 }
 
+struct GenericTraversal<'a> {
+    tree: &'a HeaviestSubtreeForkChoice,
+    pending: Vec<Slot>,
+}
+
+impl<'a> GenericTraversal<'a> {
+    pub fn new(tree: &'a HeaviestSubtreeForkChoice) -> Self {
+        Self {
+            tree,
+            pending: vec![tree.root().0],
+        }
+    }
+}
+
+impl<'a> Iterator for GenericTraversal<'a> {
+    type Item = Slot;
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.pending.pop();
+        if let Some(slot) = next {
+            let children: Vec<_> = self
+                .tree
+                .children(&(slot, Hash::default()))
+                .unwrap()
+                .iter()
+                .map(|(child_slot, _)| *child_slot)
+                .collect();
+            self.pending.extend(children);
+        }
+        next
+    }
+}
+
+pub fn get_unknown_last_index(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut repairs = Vec::new();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if repairs.len() >= limit {
+            break;
+        }
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.known_last_index().is_none() {
+                repairs.push(ShredRepairType::HighestShred(slot, slot_meta.received));
+                processed_slots.insert(slot);
+            }
+        }
+    }
+    repairs
+}
+
+pub fn get_closest_completion(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut v: Vec<(Slot, u64)> = Vec::default();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.is_full() {
+                continue;
+            }
+            if let Some(last_index) = slot_meta.known_last_index() {
```

Comment:
I'm wondering if we will run into a problem when

1. A slot is really close to completion, say it's only missing the last shred
2. The slot doesn't have the last index, i.e. missing the last shred
3. The slot is not being prioritized by `get_best_unknown_last_index()` because `for (_slot, tree) in self.trees.iter() {` will have to iterate through the earlier trees before getting to this particular tree

Should we also prioritize `get_best_unknown_last_index` by how close we are to finishing the slot?

Reply:
Updated this this to sort by `blockstore.get_index(slot).data().num_shreds()`